pallets/msa/src/lib.rs

```diff
@@ -1360,83 +1351,105 @@ impl<T: Config> Pallet<T> {
 		let current_block = frame_system::Pallet::<T>::block_number();
 
 		let max_lifetime = Self::mortality_block_limit(current_block);
-		if max_lifetime <= signature_expires_at {
-			Err(Error::<T>::ProofNotYetValid.into())
-		} else if current_block >= signature_expires_at {
-			Err(Error::<T>::ProofHasExpired.into())
-		} else {
-			let bucket_num = Self::bucket_for(signature_expires_at.into());
-			<PayloadSignatureBucketCount<T>>::try_mutate(
-				bucket_num,
-				|bucket_signature_count: &mut u32| -> DispatchResult {
-					let limit = T::MaxSignaturesPerBucket::get();
-					ensure!(
-						*bucket_signature_count < limit,
-						Error::<T>::SignatureRegistryLimitExceeded
-					);
-					let new_count =
-						bucket_signature_count.checked_add(1).ok_or(ArithmeticError::Overflow)?;
-					*bucket_signature_count = new_count;
-
-					// Tests for noop fail with "storage has been mutated"
-					// if this try_mutate does not live inside this block
-					<PayloadSignatureRegistry<T>>::try_mutate(
-						bucket_num,
-						signature,
-						|maybe_mortality_block| -> DispatchResult {
+		ensure!(max_lifetime > signature_expires_at, Error::<T>::ProofNotYetValid);
+		ensure!(current_block < signature_expires_at, Error::<T>::ProofHasExpired);
+
+		// Make sure it is not in the registry
+		ensure!(
+			!<PayloadSignatureRegistryRing<T>>::contains_key(signature),
+			Error::<T>::SignatureAlreadySubmitted
+		);
+
+		// Get the current pointer, or if this is the initialization, generate an empty pointer
+		let pointer =
+			PayloadSignatureRegistryRingPointer::<T>::get().unwrap_or(SignatureRegistryPointer {
+				head: signature.clone(),
+				tail: signature.clone(),
+				count: 0,
+			});
+
+		// Default to the current tail in case we are still filling up
+		let mut new_tail: MultiSignature = pointer.tail.clone();
+
+		// We are now wanting to overwrite prior signatures
+		// Do we need this > 0?
+		let is_ring_buffer_full: bool =
+			pointer.count > 0 && pointer.count == T::MaxSignaturesStored::get().unwrap_or(0);
```

Comment:
You are right. It is not optional. I'll remove and that removes any possible need for the pointer.count > 0.

Reply:
So it turns out that it is an optional because it is used as the max for the StorageMap max and that requires it to implement `Get<Option<u32>>`.

I still removed the pointer.count > 0 as we should not be adding signatures if the MaxSignaturesStored is unset or set to 0.